presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperator.java

```diff
@@ -0,0 +1,306 @@
+        checkArgument(unnestChannels.size() == unnestTypes.size(), "unnest channels or types has wrong size");
+        this.unnesters = unnestTypes.stream()
+                .map((Type nestedType) -> createUnnester(nestedType, isLegacyUnnest))
+                .collect(toImmutableList());
+        this.unnestOutputChannelCount = unnesters.stream().mapToInt(Unnester::getChannelCount).sum();
+
+        this.withOrdinality = withOrdinality;
+
+        this.outputChannelCount = unnestOutputChannelCount + replicateTypes.size() + (withOrdinality ? 1 : 0);
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public void finish()
+    {
+        finishing = true;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finishing && currentPage == null;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finishing && currentPage == null;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkState(!finishing, "Operator is already finishing");
+        requireNonNull(page, "page is null");
+        checkState(currentPage == null, "currentPage is not null");
+
+        currentPage = page;
+        currentPosition = 0;
+        resetBlockBuilders();
+    }
+
+    private void resetBlockBuilders()
+    {
+        for (int i = 0; i < replicateTypes.size(); i++) {
+            Block newInputBlock = currentPage.getBlock(replicateChannels.get(i));
+            replicatedBlockBuilders.get(i).resetInputBlock(newInputBlock);
+        }
+
+        for (int i = 0; i < unnestTypes.size(); i++) {
+            int inputChannel = unnestChannels.get(i);
+            Block unnestChannelInputBlock = currentPage.getBlock(inputChannel);
+            unnesters.get(i).resetInput(unnestChannelInputBlock);
+        }
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (currentPage == null) {
+            return null;
+        }
+
+        PageBuilderStatus pageBuilderStatus = new PageBuilderStatus(MAX_BYTES_PER_PAGE);
+        prepareForNewOutput(pageBuilderStatus);
+
+        int outputRowCount = 0;
+
+        while (currentPosition < currentPage.getPositionCount()) {
+            outputRowCount += processCurrentPosition();
+            currentPosition++;
+
+            if (outputRowCount >= MAX_ROWS_PER_BLOCK || pageBuilderStatus.isFull()) {
+                break;
+            }
+        }
+
+        Block[] outputBlocks = buildOutputBlocks();
+
+        if (currentPosition == currentPage.getPositionCount()) {
+            currentPage = null;
+            currentPosition = 0;
+        }
+
+        return new Page(outputBlocks);
+    }
+
+    private int processCurrentPosition()
+    {
+        // Determine number of output rows for this input position
+        int maxEntries = getCurrentMaxEntries();
```

Comment:
Nit: This function only contains one line and it's only used here. We can inline it.